public/models/interfaces.ts

```diff
@@ -205,6 +205,7 @@ export type Detector = {
   taskState?: DETECTOR_STATE;
   taskProgress?: number;
   taskError?: string;
+  dataSourceId? : string;
```

Comment:
Why can't we store the selected data source ID as a redux store item, and only update when a new ID is selected? Then you don't need to pass around ad-hoc `dataSourceId`s. If we use redux middleware for all of the calls made from client-side, the persisted data source ID within that store can handle all of that and no propagating needs to happen at all from client-side.

Thinking long-term with supporting multiple data sources, I could see some places where more fine-grained IDs need to be attached to the data models, so it makes sense to add as an optional field for `Detector` interface. But even then, this could all be persisted in the redux store and those details don't need to be handled from client-side components at all.

Reply:
Thanks for the suggestion to utilize the redux store for managing the dataSourceId. I understand the benefits of centralizing the state in Redux, especially for consistency and potentially reducing the complexity of propagating the IDs through components.

However, I have some concerns about relying entirely on the Redux store for this purpose. In the future when our application architecture supports multiple active data sources in different components simultaneously, which might lead to complications if we centralize the dataSourceId state. This approach could limit the flexibility needed for some components to operate independently under different contexts, which we anticipate will be a requirement as we scale and introduce more complex features.

Additionally, managing fine-grained control over data sources at the component level allows us to more easily integrate with other parts of our infrastructure that are not as tightly coupled with Redux. It also provides clearer traceability for data flow in complex workflows, which can aid in debugging and maintenance. Our application already have an overloaded redux state and it even crashed the Redux Debug DevTool because the state is too long to parse on some big pages like detector details page.

Reply:
> This approach could limit the flexibility needed for some components to operate independently under different contexts

Could you explain how this limits it? I'm not saying we never should have component-level context for different subsets of data source IDs, for example, but rather, by persisting the selected data source ID(s) from the top-down menu in some centralized store, we immediately aren't bound/forced to propagate those top-level IDs as props through many components that may not need it at all, but have to pass through to its children.

From my understanding, the current and near-future approach is to start by selecting a single or multiple data source IDs from a top-level picker. My proposal is that by persisting this general top-level data in redux store, we eliminate all pass-through props and data model changes needed to persist those ID(s) on their own.

Especially for the initial release which is bound to at most one data source ID, this completely eliminates the need to pass any props at all. As things could become more complex, we can introduce more IDs & props where appropriate if local, component-level context is needed, for example.

Reply:
Lmk your thoughts on this too, if you prefer some way based on the current and future plans of MDS and state management @huyaboo @BionIT

Reply:
let's allow the Neo team to decide if centralized state management suits the neo goal and could benefit the future Neo roadmap. This should be an implementation requirement for all plugins, if necessary. 

Regardless, I considered storing this ID in a Redux store and discussed this solution with Lu initially. The reason I didn't proceed is based on my current experience with adding MDS support for AD, and these concers:
- **Increased complexity**: Implementing Redux adds layers of complexity to the application. And redux often requires writing more boilerplate code, for example I have to setup actions, reducers, selectors, and middleware to manage just a simple ID. 
- **Performance overhead**: Redux connects components to the global store, which can lead to unnecessary re-renders if not carefully optimized. This is already happening to AD, almost all pages gets re-rendered multiple times when opening it.
- **Debugging difficulty**: Debugging Redux can be more challenging than simple prop passing. To understand how and why the dataSourceId is changing, we have to rely on Redux DevTools, as opposed to simply looking at component props. However, as I mentioned earlier regarding the performance overhead, Redux DevTools crashes on some AD pages because the state is too long to serialize. This leaves us with no mechanism to effectively debug the issue. 

Therefore, the reason I choose to make dataSourceId a prop include simplicity, less overhead, easier debugging and testing, reduced coupling. I do agree that when it's used by many components at different nesting levels, a Redux store is definitely the better choice. However, I believe that only the detector detail page has a really nested structure. Additionally, the dataSource picker is disabled on the detector detail page, so there's no need to pass dataSourceId around there. For the create detector page, we only allow customers choose it in the first step. It becomes a read-only value in subsequent steps, so there are no nesting concerns on the create page.

Reply:
> Increased complexity: Implementing Redux adds layers of complexity to the application. And redux often requires writing more boilerplate code, for example I have to setup actions, reducers, selectors, and middleware to manage just a simple ID.

When done correctly, it actually simplifies the complexity, since it reduces a lot of persisted/shared/common state management that otherwise needs to be handled in the individual components and passed down as props. It's not a lot of boilerplate to add this state - for example, we persist just the `requesting` and `errorMessage` vars for creating sample data [here](https://github.com/opensearch-project/anomaly-detection-dashboards-plugin/blob/main/public/redux/reducers/sampleData.ts). We could do basically the same thing, except it would be even simpler, and we only need to persist the id. We don't even need to make any server-side calls. Then, it's just a one line change to add the reducer as part of the global reducer.

> Performance overhead: Redux connects components to the global store, which can lead to unnecessary re-renders if not carefully optimized. This is already happening to AD, almost all pages gets re-rendered multiple times when opening it.

It is by design that there is multiple re-renders, they are not unnecessary. The elegant part about redux store is while many different calls may be happening on the server side, the client-side components automatically update when these values change, such as loading/requesting state, the total set of detectors, fetched index mappings, etc. Because we have fine-grained state, this leads to many re-renders, yes. Additionally, adding one field to subscribe to the data source ID wouldn't re-render anymore than what would be done via component-managed state.

> Debugging difficulty: Debugging Redux can be more challenging than simple prop passing. To understand how and why the dataSourceId is changing, we have to rely on Redux DevTools, as opposed to simply looking at component props. However, as I mentioned earlier regarding the performance overhead, Redux DevTools crashes on some AD pages because the state is too long to serialize. This leaves us with no mechanism to effectively debug the issue.

Can you explain what type of debugging is needed exactly? Persisting an ID would be very straightforward - at the top-level of pages, set the data source ID from the URL into redux store once. Now, any component can simply read that ID from store at any point - it would only be removed when client-side store is wiped out (refreshing the page). It's even simpler in that there is no server-side calls being made and nothing else needs to be tracked.

I understand how some of the complexity comes about, where the code is not totally consistent in managing some state at component-level, and some at redux level. But adding a single central state for this ID will simplify things a lot and reduce 100% of props passing. It can be equally useful when this plugin expands to support multiple datasources at once, and will lead to a lot less code changes.

I don't see the current approach as scalable, and we should avoid needing to find workaround solutions like needing to append `dataSourceId` to the `DetectorListItem` interface, for example. We can set up the small amount of infrastructure now and it will make things much simpler as MDS evolves and our plugin integrates more with it.

Reply:
+1 to @ohltyler suggestion here to add dataSourceId to redux store. The reason being, it would be simple to pass dataSourceId to various components. And dataSourceId would be the just first param for the store and later on we can have more. If not now, can we create an issue to refactor this later.

Reply:
I'm hesitant to have an issue and refactor later. Reason being, we don't want to add a field in a large number of places with plans to immediately remove them. This can lead to hard-to-find bugs if some places are missed, NPE issues, and the general difficulty of if those components or data models are updated after this change, but before it is removed. It could also potentially end up in ui metadata fields at some point on cluster-side and lead to BWC parsing issues.

Reply:
If we really dislike the idea of passing `dataSourceId` as a prop, I can change it to be read from the URL. There are two possible workflows (listed below). Regardless of which one we encounter, reading `dataSourceId` from the Redux store is completely unnecessary and over-engineered.

Workflow 1: When `dataSourceId` is not present in the URL:

1. We first render the page with a dataSource picker and get the default `dataSourceId` from the dataSource picker.
2. We update the URL with this `dataSourceId`.
3. (If we want to store this ID in the Redux store) We update the Redux store with this `dataSourceId`.


Workflow 2: When `dataSourceId` is present in the URL (e.g., a customer bookmarked the page):

1. We render the page with the given URL, which means we already have a `dataSourceId`.
2. We render the dataSource picker and make all the API calls with this `dataSourceId` to render the page.

In these two workflows, reading `dataSourceId` from the Redux store is completely needless. It doesn't serve as a source of truth because the URL is the only source of truth. We render the page based on what is present in the URL, so why waste memory storing this ID in the Redux store, which interacts with 5+ reducers?

Reply:
My $0.02:

From data source side, there are certain reasons to have state management:
1. **We recommend the use of states to reduce network calls for certain components.** For instance, if a `DataSourceView` component is repeatedly being used for different pages, it is recommended to keep both the `id` and `label` values and pass them into the component. This saves the component a network call and potentially a re-render of the page. Now this would require setting the state of both an `id` and a `label` and this might not be relevant to AD since the `DataSourceView` is used only in the detector detail page.
2. **At the plugin-level, how deeply/how many times is `dataSourceId` being passed in?** If it's being referenced frequently, having to check for `dataSourceId` being passed into every component is more tedious than simply fetching from the redux store.
3.  As for a state management requirement, this should be **case-by-case**. Several plugins only have one/a few pages to pass around the `dataSourceId`. It would be unreasonable to require state management for a prop that's being passed around a handful of times.

Reply:
Something that would be really cool is having some way to persist the IDs and make them accessible for all of the redux fns that make server-side calls that end up making cluster calls. Then all of the redux calls made on client-side wouldn't even need to pass them through; the middleware could pick up the persisted ID and pass it directly

Reply:
> Please avoid saying this; using redux to maintain common/global/shared state is a popular, practical, and clean solution.

I agree redux is a good library, I'm not saying it's useless :) Just the reason I think it's unnecessary is because that the dataSourceId in the URL is our primary source of truth for identifying a current resource or view, using it directly can simplify our state management. 

If we are retrieving the dataSourceId from the URL on every different page, why not just use it directly instead of following a cycle of use -> update redux store -> read from redux store -> use? For every child component on each page, we could either read the dataSourceId from the Redux store that was maintained on the previous parent page, or simply read the dataSourceId directly from the previous parent page. If MDS adds more attributes in the future, these will also be in the URL. Therefore, I believe that no approach is inherently better than another; they are just different ways of achieving the same outcome. 

The reason I didn't choose the Redux store approach when adding MDS support on the Overview page is due to an existing bug that caused an infinite loop in the global Redux store state changes. This took up more of my time than usual for debugging because you need to understand how every global state is changing behind the scenes. 

Regardless of the approach, we need to set the dataSourceId in DetectorListItem due to this being an interface change. The only difference is whether we set it with the value read from the URL or the value obtained from the Redux store.

I can make the change if this is really a blocker. Guess it's time to disagree and commit :)

Reply:
Maybe there's still confusion. I'm not saying we don't use it directly - I'm saying that by persisting it, we don't need to have any props passed down to any component. See previous comment:
> The whole point of having a redux store is it can be set once (either from url or from the data source picker), and then is available for any component to use, with no extra state or variables needing to be managed or passed around, at all

The idea is it gets set only the few times at the top-level of pages where its parsing from the URL or from the data source picker. Then any component throughout the entire client-side of the plugin can use that URL - no props needed or needing to find workarounds to propagate the ID through child components. The best benefit is it reduces all of that code and makes it much more flexible, should we need to persist more or different data related to MDS in the future without needing to update interfaces or prop data types.